doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+* `dds_ktopic`
+  * stores pointer to `ddsi_topic_definition`
+* Has built-in DCPSTopic topic and contains
+  * Topic name
+  * Type name
+  * Topic key
+  * QoS settings
+* `dds_find_topic_locally` w/o SDEP
+* `dds_find_topic_globally` with SDEP
+
+#### Static discovery
+
+* Static discovery shall be supported
+  * Having RouDi read in a config file, which has
+    * All the topics listed
+    * Wiring pub/sub information
+  * Static shared memory with static number of pub/subs inside
+  * During init phase pub/sub are connected
+* Create abstract interface for `ServiceRegistry`
+  * Add new class to `StaticServiceRegistry::StaticServiceRegistry(std::map<CaproIdString_t, instance_t> map)`
+    * Notification-based callback is only called once and and user gets the complete service registry
+    * Polling with `FindService()` will always return the same result
+  * Alternative would be a `finalize()` method after init phase
+
+#### Sychronous, one-shot RPC (Polling)
+
+##### Alternative A: IPC channel usage
+
+Stick with IPC channel
+
+Pro:
+
+* No code changes needed
+* More memory efficient
+* No managment overhead on iceoryx-side
+
+Con:
+
+* No userspace solution
+  * IPC channel uses OS functions during runtime (switch between kernel and user space)
+* Upper frame limit may lead to segregation into several frames
+
+##### Alternative B: Request/ response feature usage
+
+* Re-use [request/ response feature](https://github.com/eclipse-iceoryx/iceoryx/issues/27)
+
+Pro:
+
+* Re-use building blocks to avoid code duplication
+* [Dogfooding](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) your own features makes you aware of
+potential issues
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared to a simple RPC call
+
+#### Event-based notification
+
+##### Alternative A: Shared memory + IPC channel
+
+* Send ID's over IPC channel and read ABA-counter/ServiceDiscoveryChangeCounter from shared memory
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it (read-only)
+
+Pro:
+
+* Fast data transfer
+* Could replace `InterfacePort`s
+
+Contra:
+
+* Bespoken solution more complex than todays approach (thread-safety required, lock-free mechanism needed)
+* In an optimal case IPC channel communication shall only be used for startup
+
+##### Alternative B: Built-in topic based on `InterfacePort`'s
+
+* Built-in topic approach based on `InterfacePort`'s
+  * Sniff and intercept `CaproMessage`
+* Also see
+  * [overview diagram](diagrams/service-discovery-sequence-diagram.puml)
+  * [sequence diagram](diagrams/service-discovery-overview.puml)
+
+Pro:
+
+* Close to DDS mechanism
+* `InterfacePort` already available
+* Synergies with gateway mechanisms
+* Gateways can benefit from event notification mechanism
+  * No need for a polling-loop anymore
+
+Con:
+
+* Overhead, `Listner` will wake up on every `CaproMessage`
```

Comment:
I suspect even in the dynamic case there is little change after a startup phase, otherwise you have a highly artificial systems where services come and go often (unlikely). Hence I think this overhead is tolerable.